crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1908 @@
+ * 32-byte seeds public "rho", used to generate the matrix, and private "sigma",
+ * used to generate the secret vector |s|.
+ *
+ * The second random input |z| is copied verbatim into the Fujisaki-Okamoto
+ * (FO) transform "implicit-rejection" secret (the |z| component of the private
+ * key), which thwarts chosen-ciphertext attacks, provided decap() runs in
+ * constant time, with no side channel leaks, on all well-formed (valid length,
+ * and correctly encoded) ciphertext inputs.
+ */
+static __owur
+int genkey(const uint8_t d[ML_KEM_RANDOM_BYTES],
+           const uint8_t z[ML_KEM_RANDOM_BYTES],
+           scalar *tmp, EVP_MD_CTX *mdctx,
+           uint8_t *pubenc, ML_KEM_KEY *key)
+{
+    uint8_t hashed[2 * ML_KEM_RANDOM_BYTES];
+    const uint8_t *const sigma = hashed + ML_KEM_RANDOM_BYTES;
+    uint8_t augmented_seed[ML_KEM_RANDOM_BYTES + 1];
+    const ML_KEM_VINFO *vinfo = key->vinfo;
+    CBD_FUNC cbd_1 = cbd1[vinfo->variant];
+    int rank = vinfo->rank;
+    uint8_t counter = 0;
+
+    /*
+     * Use the "d" seed salted with the rank to derive the public and private
+     * seeds rho and sigma.
+     */
+    memcpy(augmented_seed, d, ML_KEM_RANDOM_BYTES);
+    augmented_seed[ML_KEM_RANDOM_BYTES] = (uint8_t) rank;
+    if (!hash_g(hashed, augmented_seed, sizeof(augmented_seed), mdctx, key))
+        return 0;
+    memcpy(key->rho, hashed, ML_KEM_RANDOM_BYTES);
+    if (!matrix_expand(mdctx, key)
+        || !gencbd_vector(key->s, cbd_1, &counter, sigma, rank, mdctx, key))
+        return 0;
+    vector_ntt(key->s, rank);
+    /* FIPS 203 |e| vector */
+    if (!gencbd_vector(tmp, cbd_1, &counter, sigma, rank, mdctx, key))
+        return 0;
+    vector_ntt(tmp, rank);
+
+    /* Fill in the public key */
+    matrix_mult_transpose(key->t, key->m, key->s, rank);
+    vector_add(key->t, tmp, rank);
+    encode_pubkey(pubenc, key);
+    if (!hash_h(key->pkhash, pubenc, vinfo->pubkey_bytes, mdctx, key))
+        return 0;
+
+    /* Save "z" portion of seed for "implicit rejection" on failure */
+    memcpy(key->z, z, ML_KEM_RANDOM_BYTES);
+    return 1;
+}
+
+/*-
+ * FIPS 203, Section 6.2, Algorithm 17: "ML-KEM.Encaps_internal".
+ * This is the deterministic version with randomness supplied externally.
+ *
+ * The caller must pass space for two vectors in |tmp|.
+ * The |ctext| buffer have space for the ciphertext of the ML-KEM variant
+ * of the provided key.
+ */
+static
+int encap(uint8_t *ctext, uint8_t secret[ML_KEM_SHARED_SECRET_BYTES],
+          const uint8_t entropy[ML_KEM_RANDOM_BYTES],
+          scalar *tmp, EVP_MD_CTX *mdctx, const ML_KEM_KEY *key)
+{
+    uint8_t input[ML_KEM_RANDOM_BYTES + ML_KEM_PKHASH_BYTES];
+    uint8_t Kr[ML_KEM_SHARED_SECRET_BYTES + ML_KEM_RANDOM_BYTES];
+    uint8_t *r = Kr + ML_KEM_SHARED_SECRET_BYTES;
+
+#if ML_KEM_SEED_BYTES != ML_KEM_SHARED_SECRET_BYTES + ML_KEM_RANDOM_BYTES
+# error "ML-KEM keygen seed length != shared secret + random bytes length"
```

Comment:
There must be a better place for this compile time check…
Near the top of the file seems reasonable.

Likewise for other instances of similar below.

Reply:
Done.